File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        if(root == null) {
            return 0;
        }

        int[] res = new int[1];
        helper(root, Integer.MIN_VALUE, res);
        return res[0];
    }

    public void helper(TreeNode root, int maxTillNow, int[] res) {
        if(root == null) {
            return;
        }

        if(root.val >= maxTillNow) res[0]++;
        maxTillNow = Math.max(maxTillNow, root.val);

        helper(root.left, maxTillNow, res);
        helper(root.right, maxTillNow, res);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;

        int res = 0;
        helper(root, res);
        return res;
    }


    void helper(TreeNode* root, int& res) {
        if(!root) return;

        res = max(res, maxDepth(root->left) + maxDepth(root->right));
        helper(root->left, res);
        helper(root->right, res);
    }

    int maxDepth(TreeNode* root) {
        if(!root) return 0;

        return 1 + max(maxDepth(root->left), maxDepth(root->right));
    }
};
